File: algorithms/algorithms-cplusplus/leetcode/Question_0911_Online_Election.cc
#include <vector>
#include <unordered_map>

using namespace std;

class TopVotedCandidate {
 public:
  TopVotedCandidate(vector<int>& persons, vector<int>& times) {
    int n = persons.size(), lead = -1;
    this->times = times;
    unordered_map<int, int> count;
    for (int i = 0; i < n; ++i) {
      lead = ++count[persons[i]] >= count[lead] ? persons[i] : lead;
      m[times[i]] = lead;
    }
  }
  
  int q(int t) {
    return m[*--upper_bound(times.begin(), times.end(), t)];
  }

 private:
  unordered_map<int, int> m;
  vector<int> times;
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0911_Online_Election.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Question_0911_Online_Election {
    private Map<Integer, Integer> m;
    private int[] time;

    public Question_0911_Online_Election(int[] persons, int[] times) {
        this.m = new HashMap<>();
        int n = persons.length, lead = -1;
        Map<Integer, Integer> count = new HashMap<>();
        time = times;
        for (int i = 0; i < n; ++i) {
            count.put(persons[i], count.getOrDefault(persons[i], 0) + 1);
            if (i == 0 || count.get(persons[i]) >= count.get(lead)) {
                lead = persons[i];
            }
            m.put(times[i], lead);
        }
    }

    public int q(int t) {
        int i = Arrays.binarySearch(time, t);
        return i < 0 ? m.get(time[-i - 2]) : m.get(time[i]);
    }
}
